File: lab12/prac/lec24/main.py
'''
Collection of efficient string algorithms useful for programming contests.
Functions included:
  - prefix_function(s)
  - kmp_search(text, pattern)
  - aho_corasick(patterns)
  - suffix_array(s)
  - lcp_array(s, sa)
Usage examples at the bottom.
'''
from collections import deque


def prefix_function(s: str) -> list[int]:
    """
    Computes the prefix function (pi array) for string s.
    pi[i] = length of the longest proper prefix of s[:i+1] which is also a suffix.
    Time: O(n)
    """
    n = len(s)
    pi = [0] * n
    j = 0
    for i in range(1, n):
        while j > 0 and s[i] != s[j]:
            j = pi[j - 1]
        if s[i] == s[j]:
            j += 1
        pi[i] = j
    return pi


def kmp_search(text: str, pattern: str) -> list[int]:
    """
    Finds all occurrences of pattern in text using KMP.
    Returns list of starting indices (0-based) in text where pattern matches.
    Time: O(n + m)
    """
    if not pattern:
        return list(range(len(text) + 1))
    pi = prefix_function(pattern)
    res = []
    j = 0
    for i, ch in enumerate(text):
        while j > 0 and ch != pattern[j]:
            j = pi[j - 1]
        if ch == pattern[j]:
            j += 1
            if j == len(pattern):
                res.append(i - j + 1)
                j = pi[j - 1]
    return res


class AhoCorasick:
    """
    Aho–Corasick automaton for multi-pattern string matching.
    Build with a list of patterns, then search in text.
    Time: Build O(sum len), Search O(text length + matches).
    """
    def __init__(self, patterns: list[str]):
        self.num_nodes = 1
        self.edges = []           # List of dicts: edges[u][c] = v
        self.fail = []            # failure links
        self.output = []          # output[u] = list of pattern indices ending at u
        for _ in range(1):
            self.edges.append({})
            self.fail.append(0)
            self.output.append([])
        # insert patterns
        for idx, pat in enumerate(patterns):
            u = 0
            for ch in pat:
                if ch not in self.edges[u]:
                    self.edges[u][ch] = self.num_nodes
                    self.edges.append({})
                    self.fail.append(0)
                    self.output.append([])
                    self.num_nodes += 1
                u = self.edges[u][ch]
            self.output[u].append(idx)
        # build failure links
        q = deque()
        for ch, v in self.edges[0].items():
            q.append(v)
            self.fail[v] = 0
        while q:
            u = q.popleft()
            for ch, v in self.edges[u].items():
                q.append(v)
                f = self.fail[u]
                while f > 0 and ch not in self.edges[f]:
                    f = self.fail[f]
                self.fail[v] = self.edges[f].get(ch, 0)
                self.output[v] += self.output[self.fail[v]]

    def search(self, text: str) -> list[tuple[int, int]]:
        """
        Search text for the patterns.
        Returns list of (position, pattern_index) pairs where a pattern ends at position.
        Position is the index in text of the last character of the match.
        """
        u = 0
        results = []
        for i, ch in enumerate(text):
            while u > 0 and ch not in self.edges[u]:
                u = self.fail[u]
            u = self.edges[u].get(ch, 0)
            for pat_idx in self.output[u]:
                results.append((i, pat_idx))
        return results


def suffix_array(s: str) -> list[int]:
    """
    Constructs the suffix array of s. Returns list of starting indices of suffixes in sorted order.
    Time: O(n log n)
    """
    s += chr(0)  # append sentinel
    n = len(s)
    k = 0
    sa = list(range(n))
    rank_arr = [ord(c) for c in s]
    tmp = [0] * n
    def cmp(a, b):
        if rank_arr[a] != rank_arr[b]:
            return rank_arr[a] - rank_arr[b]
        ra = rank_arr[a + k] if a + k < n else -1
        rb = rank_arr[b + k] if b + k < n else -1
        return ra - rb
    while (1 << k) < n:
        sa.sort(key=lambda x: (rank_arr[x], rank_arr[x + k] if x + k < n else -1))
        tmp[sa[0]] = 0
        for i in range(1, n):
            tmp[sa[i]] = tmp[sa[i-1]] + (cmp(sa[i-1], sa[i]) < 0)
        rank_arr, tmp = tmp, rank_arr
        k += 1

    return sa[1:]  # drop sentinel index


def inverse_sa(sa: list[int]) -> list[int]:
    """
    Compute inverse of suffix array: inv_sa[i] = position of suffix i in SA.
    """
    n = len(sa)
    inv = [0]*n
    for i, si in enumerate(sa):
        inv[si] = i
    return inv


def lcp_array(s: str, sa: list[int]) -> list[int]:
    """
    Constructs the LCP array using Kasai's algorithm.
    lcp[i] = LCP(sa[i], sa[i+1]).
    Time: O(n)
    """
    n = len(s)
    rank_arr = [0] * n
    for i, idx in enumerate(sa):
        rank_arr[idx] = i
    h = 0
    lcp = [0] * (n - 1)
    for i in range(n):
        if rank_arr[i] == 0:
            continue
        j = sa[rank_arr[i] - 1]
        while i + h < n and j + h < n and s[i + h] == s[j + h]:
            h += 1
        lcp[rank_arr[i] - 1] = h
        if h > 0:
            h -= 1
    return lcp


def lcp_query(i: int, j: int, sa: list[int], lcp: list[int], inv: list[int]) -> int:
    """
    Compute LCP of suffixes at indices i, j using RMQ over lcp array (naive).  
    """
    ri, rj = inv[i], inv[j]
    if ri > rj:
        ri, rj = rj, ri
    return min(lcp[ri:rj]) if ri!=rj else len(sa)-i


def build_nextmatch(p: str, alphabet: set[str]) -> list[dict[str,int]]:
    """
    Build the next-match table for p over given alphabet.
    nextmatch[i][c] = new state after reading c from state i.
    Time: O(m * |alphabet|)
    """
    m = len(p)
    pi = prefix_function(p)
    # allocate table
    nextmatch = [ {c:0 for c in alphabet} for _ in range(m+1) ]
    # state 0
    for c in alphabet:
        nextmatch[0][c] = 1 if m>0 and p[0] == c else 0
    k = 0
    for i in range(1, m+1):
        if i < m:
            for c in alphabet:
                if p[i] == c:
                    nextmatch[i][c] = i + 1
                else:
                    nextmatch[i][c] = nextmatch[pi[i-1]][c]
            k = nextmatch[k][p[i-1]]
    return nextmatch


def fsm_search(t: str, p: str) -> bool:
    """
    Check if p is substring of t via FSM matching.
    """
    if not p:
        return True
    alphabet = set(t) | set(p)
    nextmatch = build_nextmatch(p, alphabet)
    state = 0
    m = len(p)
    for c in t:
        state = nextmatch[state].get(c, 0)
        if state == m:
            return True
    return False

# Example usage
if __name__ == '__main__':
    text = 'ababcababc'
    pat = 'ababc'
    print('KMP matches at:', kmp_search(text, pat))
    patterns = ['he', 'she', 'hers', 'his']
    ac = AhoCorasick(patterns)
    print('AC matches:', ac.search('ahishers'))
    sa = suffix_array('banana')
    print('SA of banana:', sa)
    print('LCP of banana:', lcp_array('banana', sa))


    # Sample text and pattern
    text = "ababcababa"
    pat = "ababa"

    print("Pattern:", pat)
    print("Prefix function:", prefix_function(pat))
    print("KMP search positions:", kmp_search(text, pat))
    print("FSM match found?", fsm_search(text, pat))

    # Suffix array and LCP for a sample string
    sample = "banana"
    sa = suffix_array(sample)
    inv = inverse_sa(sa)
    lcp_arr = lcp_array(sample, sa)

    print("\nString:", sample)
    print("Suffix Array:", sa)
    print("LCP Array:", lcp_arr)
    # LCP of suffixes starting at positions 1 ('anana') and 3 ('ana')
    print("LCP(1,3):", lcp_query(1, 3, sa, lcp_arr, inv))

File: lab12/prac/lec24/main.c
/*
 * Collection of efficient string algorithms useful for programming contests.
 * Functions included:
 *   - prefix_function(s, n, pi)
 *   - kmp_search(text, n, pattern, m, res, &res_count)
 *   - AhoCorasick automaton
 *   - suffix_array(s, n, sa)
 *   - lcp_array(s, n, sa, lcp)
 *   - lcp_query(i, j, sa, inv, lcp, n)
 *   - fsm_search(text, n, pattern, m)
 * Usage examples in main().
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

/* Prefix function */
void prefix_function(const char *s, int n, int *pi) {
    pi[0] = 0;
    int j = 0;
    for (int i = 1; i < n; i++) {
        while (j > 0 && s[i] != s[j]) j = pi[j-1];
        if (s[i] == s[j]) j++;
        pi[i] = j;
    }
}

/* KMP search: returns dynamic array of positions, sets res_count */
int *kmp_search(const char *text, int n, const char *pat, int m, int *res_count) {
    int *res = malloc((n+1) * sizeof(int));
    *res_count = 0;
    if (m == 0) {
        for (int i = 0; i <= n; i++) res[(*res_count)++] = i;
        return res;
    }
    int *pi = malloc(m * sizeof(int));
    prefix_function(pat, m, pi);
    int j = 0;
    for (int i = 0; i < n; i++) {
        while (j > 0 && text[i] != pat[j]) j = pi[j-1];
        if (text[i] == pat[j]) j++;
        if (j == m) {
            res[(*res_count)++] = i - m + 1;
            j = pi[j-1];
        }
    }
    free(pi);
    return res;
}

/* Aho-Corasick automaton (lowercase a-z) */
#define ALPHA 26

typedef struct ACNode {
    int next[ALPHA];
    int link;
    int *out;
    int out_size;
} ACNode;

typedef struct AhoCorasick {
    ACNode *nodes;
    int size;
} AhoCorasick;

AhoCorasick *ac_new(int max_nodes) {
    AhoCorasick *ac = malloc(sizeof(*ac));
    ac->nodes = calloc(max_nodes, sizeof(ACNode));
    ac->size = 1;
    for (int c = 0; c < ALPHA; c++) ac->nodes[0].next[c] = -1;
    ac->nodes[0].link = 0;
    ac->nodes[0].out = NULL;
    ac->nodes[0].out_size = 0;
    return ac;
}

void ac_insert(AhoCorasick *ac, const char *pat, int idx) {
    int u = 0;
    for (; *pat; pat++) {
        int c = *pat - 'a';
        if (ac->nodes[u].next[c] == -1) {
            int v = ac->size++;
            ac->nodes[u].next[c] = v;
            for (int k = 0; k < ALPHA; k++) ac->nodes[v].next[k] = -1;
            ac->nodes[v].link = 0;
            ac->nodes[v].out = NULL;
            ac->nodes[v].out_size = 0;
        }
        u = ac->nodes[u].next[c];
    }
    ac->nodes[u].out = realloc(ac->nodes[u].out, (ac->nodes[u].out_size+1)*sizeof(int));
    ac->nodes[u].out[ac->nodes[u].out_size++] = idx;
}

void ac_build(AhoCorasick *ac) {
    int *q = malloc(ac->size * sizeof(int));
    int front = 0, back = 0;
    for (int c = 0; c < ALPHA; c++) {
        int v = ac->nodes[0].next[c];
        if (v != -1) {
            ac->nodes[v].link = 0;
            q[back++] = v;
        } else {
            ac->nodes[0].next[c] = 0;
        }
    }
    while (front < back) {
        int u = q[front++];
        for (int c = 0; c < ALPHA; c++) {
            int v = ac->nodes[u].next[c];
            if (v != -1) {
                int f = ac->nodes[u].link;
                ac->nodes[v].link = ac->nodes[f].next[c];
                int link = ac->nodes[v].link;
                if (ac->nodes[link].out_size) {
                    int sz = ac->nodes[link].out_size;
                    ac->nodes[v].out = realloc(ac->nodes[v].out, (ac->nodes[v].out_size+sz)*sizeof(int));
                    memcpy(ac->nodes[v].out + ac->nodes[v].out_size, ac->nodes[link].out, sz*sizeof(int));
                    ac->nodes[v].out_size += sz;
                }
                q[back++] = v;
            } else {
                ac->nodes[u].next[c] = ac->nodes[ac->nodes[u].link].next[c];
            }
        }
    }
    free(q);
}

int *ac_search(AhoCorasick *ac, const char *text, int n, int *out_count) {
    int *out = malloc(n * 2 * sizeof(int));
    *out_count = 0;
    int u = 0;
    for (int i = 0; i < n; i++) {
        int c = text[i] - 'a';
        u = ac->nodes[u].next[c];
        for (int k = 0; k < ac->nodes[u].out_size; k++) {
            out[(*out_count)*2] = i;
            out[(*out_count)*2+1] = ac->nodes[u].out[k];
            (*out_count)++;
        }
    }
    return out;
}

/* Global variables for suffix-array comparator */
static int n_sa;
static int step_k;
static int *rank_sa;

int cmp_sa(const void *a, const void *b) {
    int i = *(const int*)a;
    int j = *(const int*)b;
    if (rank_sa[i] != rank_sa[j]) return rank_sa[i] - rank_sa[j];
    int ri = (i+step_k <= n_sa ? rank_sa[i+step_k] : -1);
    int rj = (j+step_k <= n_sa ? rank_sa[j+step_k] : -1);
    return ri - rj;
}

/* Suffix array (doubling) */
void suffix_array(const char *s, int n, int *sa) {
    n_sa = n;
    int *rk = malloc((n+1)*sizeof(int));
    int *tmp = malloc((n+1)*sizeof(int));
    for (int i = 0; i <= n; i++) {
        sa[i] = i;
        rk[i] = (i < n ? (unsigned char)s[i] : -1);
    }
    for (step_k = 1; step_k <= n; step_k <<= 1) {
        rank_sa = rk;
        qsort(sa, n+1, sizeof(int), cmp_sa);
        tmp[sa[0]] = 0;
        for (int i = 1; i <= n; i++) {
            tmp[sa[i]] = tmp[sa[i-1]] + (cmp_sa(&sa[i-1], &sa[i]) < 0);
        }
        memcpy(rk, tmp, (n+1)*sizeof(int));
    }
    /* Drop the sentinel index at position 0 */
    memmove(sa, sa+1, n * sizeof(int));
    free(rk);
    free(tmp);
}

/* LCP array (Kasai) */
void lcp_array(const char *s, int n, const int *sa, int *lcp) {
    int *rank = malloc(n*sizeof(int));
    for (int i = 0; i < n; i++) rank[sa[i]] = i;
    int h = 0;
    for (int i = 0; i < n; i++) {
        if (rank[i] > 0) {
            int j = sa[rank[i]-1];
            while (i+h < n && j+h < n && s[i+h] == s[j+h]) h++;
            lcp[rank[i]-1] = h;
            if (h) h--;
        }
    }
    free(rank);
}

int lcp_query(int i, int j, const int *sa, const int *inv, const int *lcp, int n) {
    int ri = inv[i], rj = inv[j];
    if (ri > rj) { int t=ri; ri=rj; rj=t; }
    if (ri == rj) return n - i;
    int mn = INT_MAX;
    for (int k = ri; k < rj; k++) if (lcp[k] < mn) mn = lcp[k];
    return mn;
}

/* FSM substring check */
int fsm_search(const char *text, int n, const char *p, int m) {
    if (m == 0) return 1;
    int *pi = malloc(m*sizeof(int));
    prefix_function(p, m, pi);
    int state = 0;
    for (int i = 0; i < n; i++) {
        while (state>0 && text[i] != p[state]) state = pi[state-1];
        if (text[i] == p[state]) state++;
        if (state == m) { free(pi); return 1; }
    }
    free(pi);
    return 0;
}

int main() {
    const char *text = "ababcababc";
    const char *pat  = "ababc";
    int res_count;
    int *res = kmp_search(text, strlen(text), pat, strlen(pat), &res_count);
    printf("KMP matches at:");
    for (int i = 0; i < res_count; i++) printf(" %d", res[i]);
    printf("\n");
    free(res);

    /* Aho-Corasick example */
    const char *patterns[] = {"he","she","hers","his"};
    AhoCorasick *ac = ac_new(1000);
    for (int i = 0; i < 4; i++) ac_insert(ac, patterns[i], i);
    ac_build(ac);
    int out_count;
    int *out = ac_search(ac, "ahishers", strlen("ahishers"), &out_count);
    printf("AC matches (pos, idx):");
    for (int i = 0; i < out_count; i++) printf(" (%d,%d)", out[2*i], out[2*i+1]);
    printf("\n");
    free(out);

    printf("FSM match found? %s\n",
           fsm_search("ababcababa", strlen("ababcababa"), "ababa", strlen("ababa")) ? "Yes" : "No");
    return 0;
}
